### PR TITLE
Siblings error after creating bucket

### PR DESCRIPTION
**Pull request by kellymclaughlin**
This test is a bit artificial, but it simulates a network split case that can happen with Riak, and it creates a bucket that appears to be completely unaccessible afterward.  All attempts to access the bucket seem to throw this exception in riakc_object:

```
webmachine error: path="/testtest/"
{throw,siblings,[{riakc_obj,get_metadata,1,[{file,"src/riakc_obj.erl"},{line,147}]},{riak_cs_acl,bucket_acl,2,[{file,"src/riak_cs_acl.erl"},{line,125}]},{riak_cs_acl,bucket_access,4,[{file,"src/riak_cs_acl.erl"},{line,94}]},{riak_cs_wm_bucket,check_permission,2,[{file,"src/riak_cs_wm_bucket.erl"},{line,83}]},{riak_cs_wm_bucket,forbidden,2,[{file,"src/riak_cs_wm_bucket.erl"},{line,51}]},{webmachine_resource,resource_call,3,[{file,"src/webmachine_resource.erl"},{line,166}]},{webmachine_resource,do,3,[{file,"src/webmachine_resource.erl"},{line,125}]},{webmachine_decision_core,resource_call,1,[{file,"src/webmachine_decision_core.erl"},{line,48}]}]}
```

Steps to reproduce:
1. Create a 2 node cluster.  I used these options in the `riak_core` section of `app.config`, in case it's helpful.
   {ring_creation_size, 8},
   {default_bucket_props, [{allow_mult,true}]},
2. Shut down node #2.
3. Create the bucket: `s3cmd mb s3://testtest`
4. Put a file into the bucket: `s3cmd put ~/foo s3://testtest/foo1`
   - Not clear if this is 100% necessary, but it is fun.
5. Stop Riak node #1 and start Riak node #2.
   - It is probably necessary to alter your Riak CS & Stanchion configs to use the PB port for Riak.  If so, then stop, edit, & restart CS and Stanchion
6. Create the bucket again: `s3cmd mb s3://testtest`.  The creation is successful.
7. Put a file: `s3cmd put ~/foo s3://testtest/foo2`, get the siblings error.
8. All subsequent attempts to access the bucket fail.  The `s3cmd` output says:
   
   DEBUG: Response: {'status': 500, 'headers': {'date': 'Sat, 24 Nov 2012 22:21:49 GMT', 'content-length': '170', 'content-type': 'text/html', 'server': 'MochiWeb/1.1 WebMachine/1.9.0 (someone had painted it blue)'}, 'reason': 'Internal Server Error', 'data': '<html><head><title>500 Internal Server Error</title></head><body><h1>Internal Server Error</h1>The server encountered an error while processing this request</body></html>'}
   DEBUG: S3Error: 500 (Internal Server Error)
   DEBUG: HttpHeader: date: Sat, 24 Nov 2012 22:21:49 GMT
   DEBUG: HttpHeader: content-length: 170
   DEBUG: HttpHeader: content-type: text/html
   DEBUG: HttpHeader: server: MochiWeb/1.1 WebMachine/1.9.0 (someone had painted it blue)

### CODE REVIEW

**Comment by kellymclaughlin** [comment]
I have added conflict resolution code for bucket ACLs and policies. This also includes a pretty substantial refactoring of the `bucket_access_authorize_helper` function in `riak_cs_wm_utils` and a reworking of how `riak_cs_utils:get_bucket_acl_policy` is organized.

**Comment by kellymclaughlin** [comment]
With this branch and [this](https://github.com/basho/stanchion/issues/40) `stanchion` pr, the failure scenario described above should no longer be an issue (save some minor edge cases perhaps). Attempting to create a bucket when one of the nodes is down should result in a constraint violation being caught by `stanchion` and Riak CS should return a `503 Service Unavailable` to the user. In the case that the siblings already exist, the conflict resolution logic in CS should be able to allow operations using that bucket to continue or make noise in the case of multiple owners of the same bucket.

File: src/riak_cs_acl.erl [hunk 1]
```diff
@@ -146,18 +149,57 @@ bucket_access(_, RequestedAccess, CanonicalId, RiakPid, Acl) ->
...
 bucket_acl(Bucket, RiakPid) ->
     case riak_cs_utils:check_bucket_exists(Bucket, RiakPid) of
         {ok, Obj} ->
-            %% For buckets there will not be siblings.
-            MD = riakc_obj:get_metadata(Obj),
-            MetaVals = dict:fetch(?MD_USERMETA, MD),
-            acl_from_meta(MetaVals);
+            %% For buckets there should not be siblings, but in rare
+            %% cases it may happen so check for them and attempt to
+            %% resolve if possible.
```
Comment (reiddraper) [inline src/riak_cs_acl.erl:160]: maybe put a note about purposely resolving, but _not_ writing the resolved values back to Riak?

File: src/riak_cs_wm_utils.erl [hunk 2]
```diff
@@ -398,97 +398,110 @@ extract_metadata(Headers) ->
...
+    %% No policy so emulate a policy eval failure to avoid code duplication
+    handle_policy_eval_result(Ctx#context.user, false, undefined, RD, Ctx);
+handle_acl_check_result(false, Acl, Policy, AccessType, _Deletable, RD, Ctx) ->
+    User = Ctx#context.user,
+    PolicyMod = Ctx#context.policy_module,
+    Access = PolicyMod:reqdata_to_access(RD,
+                                         AccessType,
+                                         User?RCS_USER.canonical_id),
+    PolicyResult = PolicyMod:eval(Access, Policy),
+    OwnerId = riak_cs_acl:owner_id(Acl),
```
Comment (reiddraper) [inline src/riak_cs_wm_utils.erl]: `owner_id/1` doesn't exist, just `/2`

File: src/riak_cs_s3_policy.erl [hunk 3]
```diff
@@ -197,6 +201,69 @@ log_supported_actions()->
...
+resolve_bucket_metadata(Metas, [_Val]) ->
+    Policys = [policy_from_meta(M) || M <- Metas],
+    resolve_bucket_policys(Policys);
+resolve_bucket_metadata(_Metas, _) ->
+    {error, multiple_bucket_owners}.
+
+-spec resolve_bucket_policys(list(policy_from_meta_result())) -> policy_from_meta_result().
+resolve_bucket_policys([Policy]) ->
+    Policy;
+resolve_bucket_policys(Policys) ->
```
Comment (reiddraper) [inline src/riak_cs_s3_policy.erl]: nbd, but s/Policys/Policies if you like

File: src/riak_cs_s3_policy.erl [hunk 4]
```diff
@@ -197,6 +201,69 @@ log_supported_actions()->
...
+    Policy1;
+newer_policy({ok, Policy1}, {ok, Policy2})
+  when Policy1?POLICY.creation_time >= Policy2?POLICY.creation_time ->
+    {ok, Policy1};
+newer_policy(_, Policy2) ->
+    Policy2.
+
+%% @doc Find the policy in a list of metadata values and
+%% convert it to an erlang term representation. Return
+%% `undefined' if an policy is not found.
```
Comment (reiddraper) [inline src/riak_cs_s3_policy.erl]: actually, returns `{error, policy_undefined}`

**Comment by reiddraper** [comment]
+1

**Pull request closed without merging**